File: react-java/backend/src/main/java/com/clt/hrm/infra/common/dtos/SearchBaseDto.java
package com.clt.hrm.infra.common.dtos;

import java.util.List;

import lombok.Data;
import lombok.EqualsAndHashCode;

@Data
@EqualsAndHashCode(callSuper=false)
public class SearchBaseDto extends BaseDto {
    private String searchText;

    // Dynamic Filter
    List<DynamicFilterDto> filters;
    
    // Pagination & Sorting
    private SortDto sort;
    private List<SortDto> sorts;
    private PaginationDto pagination;
}

File: nestjs/backend/src/infra/common/dtos/search-base.dto.ts
import { IsArray, IsOptional, IsString, ValidateNested } from 'class-validator';
import { Type } from 'class-transformer';
import { PaginationDto } from './pagination.dto';
import { SortDto } from './sort.dto';
import { DynamicFilterDto } from './dynamic-filter.dto';
import { BaseDto } from './base.dto';

export class SearchBaseDto extends BaseDto {
  @IsOptional()
  @IsString()
  searchText?: string;

  @IsOptional()
  @ValidateNested()
  @Type(() => PaginationDto)
  pagination?: PaginationDto;

  @IsOptional()
  @ValidateNested()
  @Type(() => SortDto)
  sort?: SortDto;

  @IsOptional()
  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => SortDto)
  sorts?: SortDto[];

  @IsOptional()
  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => DynamicFilterDto)
  filters?: DynamicFilterDto[];
}
